README.md
# Bottleneck Calculator

[![GoDoc](https://godoc.org/github.com/ssgreg/bottleneck?status.svg)](https://godoc.org/github.com/ssgreg/bottleneck)
[![Build Status](https://travis-ci.org/ssgreg/bottleneck.svg?branch=master)](https://travis-ci.org/ssgreg/bottleneck)
[![Go Report Status](https://goreportcard.com/badge/github.com/ssgreg/bottleneck)](https://goreportcard.com/report/github.com/ssgreg/bottleneck)

Package `bottleneck` allows do detect bottlenecks in your code. Package uses `atomic` to minimize impact on performance. Bottleneck info returning by `Stats` function is eventually consistent.

## Example

### Which operation takes more time

The example below proves that creating and filling an array with random numbers takes less time than sorting.

```go
    for i := 1000; i < 1000000000; i *= 10 {
        // Our bc has 2 bottleneck entries:
        // 0 - array creation and filling with random numbers
        // 1 - sorting
        bc := bottleneck.NewCalculator()

        // Will add array creation and filling time slice to the
        // bottleneck entry (0).
        bc.TimeSlice(bottleneck.Index0)

        ra := make([]int, i)
        for j := 0; j < i; j++ {
            ra[j] = rand.Int()
        }

        // Will add sorting time slice to the bottleneck entry (1).
        bc.TimeSlice(bottleneck.Index1)

        sort.Ints(ra)

        entries := bc.Stats()
        fmt.Printf("array len is %v, creating new random array takes %v, sorting takes %v\n", i, entries[0].Duration, entries[1].Duration)
    }
```

The output is:

```
array len is 1000, creating new random array takes 40.059µs, sorting takes 213.93µs
array len is 10000, creating new random array takes 455.646µs, sorting takes 2.163691ms
array len is 100000, creating new random array takes 4.297062ms, sorting takes 26.020184ms
array len is 1000000, creating new random array takes 40.96832ms, sorting takes 307.671125ms
array len is 10000000, creating new random array takes 390.894645ms, sorting takes 3.574017192s
array len is 100000000, creating new random array takes 3.911881876s, sorting takes 40.516029548s
```

### Monitoring of a goroutine execution

The example below starts two goroutines. The first one is imitating a worker with two complex jobs. The second one monitors worker's executions. It shows percentage, number of calls and total execution time per each job.

```go
func main() {
    wg := sync.WaitGroup{}
    wg.Add(1)

    // Our bc has 3 bottleneck entries:
    // 0 - auxiliary: for loop, checks if break is needed
    // 1 - complex job (1)
    // 2 - complex job (2)
    bc := bottleneck.NewCalculator()

    go func() {
        defer wg.Done()
        finish := time.Now().Add(time.Second * 50)
        for {
            if time.Now().After(finish) {
                break
            }

            // Next time slice will be added to the bottleneck with index 1.
            bc.TimeSlice(bottleneck.Index1)

            // Do complex job (1).
            time.Sleep(time.Duration(rand.Int() % 500 * 1000000))
            //

            // Next time slice will be added to the bottleneck with index 2.
            bc.TimeSlice(bottleneck.Index2)

            // Do complex job (2). It's 2 times longer than job (1).
            time.Sleep(time.Duration(rand.Int() % 1000 * 1000000))
            //

            // Next time slice will be added to the bottleneck with index 0.
            bc.TimeSlice(bottleneck.Index0)
        }
    }()

    go func() {
        // Monitor that 10 times per second prints bottleneck stats.
        for {
            time.Sleep(time.Millisecond * 100)
            bns := bc.Stats()
            fmt.Printf("j1: %0.3f%%, %v, %d calls | j2: %0.3f%%, %v, %d calls\n", bns[1].Percentage, bns[1].Duration, bns[1].CallCount, bns[2].Percentage, bns[2].Duration, bns[2].CallCount)
        }
    }()

    wg.Wait()
}
```

A fragment of output:

```
j1: 1.000%, 100.260332ms, 0 calls | j2: 0.000%, 0s, 0 calls
j1: 1.000%, 201.013773ms, 0 calls | j2: 0.000%, 0s, 0 calls
j1: 1.000%, 301.297164ms, 0 calls | j2: 0.000%, 0s, 0 calls
j1: 1.000%, 402.127131ms, 0 calls | j2: 0.000%, 0s, 0 calls
j1: 0.815%, 410.234664ms, 1 calls | j2: 0.185%, 92.8835350ms, 0 calls
j1: 0.679%, 410.234664ms, 1 calls | j2: 0.321%, 193.718235ms, 0 calls
j1: 0.582%, 410.234664ms, 1 calls | j2: 0.418%, 294.723694ms, 0 calls
j1: 0.509%, 410.234664ms, 1 calls | j2: 0.491%, 395.139638ms, 0 calls
j1: 0.453%, 410.234664ms, 1 calls | j2: 0.547%, 495.800695ms, 0 calls
j1: 0.452%, 454.419565ms, 1 calls | j2: 0.548%, 551.816673ms, 1 calls
j1: 0.502%, 555.145899ms, 1 calls | j2: 0.498%, 551.816673ms, 1 calls
j1: 0.543%, 655.362978ms, 1 calls | j2: 0.457%, 551.816673ms, 1 calls
j1: 0.559%, 731.396896ms, 2 calls | j2: 0.441%, 576.023745ms, 1 calls
j1: 0.571%, 804.403279ms, 2 calls | j2: 0.429%, 603.695591ms, 2 calls
j1: 0.600%, 905.306768ms, 2 calls | j2: 0.400%, 603.695591ms, 2 calls
j1: 0.625%, 1.005623715s, 2 calls | j2: 0.375%, 603.695591ms, 2 calls
j1: 0.647%, 1.106392719s, 2 calls | j2: 0.353%, 603.695591ms, 2 calls
j1: 0.646%, 1.169336820s, 3 calls | j2: 0.354%, 641.748760ms, 2 calls
j1: 0.612%, 1.169336820s, 3 calls | j2: 0.388%, 741.900221ms, 2 calls
j1: 0.581%, 1.169336820s, 3 calls | j2: 0.419%, 842.655883ms, 2 calls
j1: 0.563%, 1.188210892s, 3 calls | j2: 0.437%, 923.956385ms, 3 calls
j1: 0.582%, 1.289025414s, 3 calls | j2: 0.418%, 923.956385ms, 3 calls
j1: 0.601%, 1.390096321s, 3 calls | j2: 0.399%, 923.956385ms, 3 calls
j1: 0.591%, 1.428142697s, 4 calls | j2: 0.409%, 986.964089ms, 3 calls
j1: 0.574%, 1.443765872s, 4 calls | j2: 0.426%, 1.072230598s, 4 calls
j1: 0.590%, 1.544728063s, 4 calls | j2: 0.410%, 1.072230598s, 4 calls
j1: 0.605%, 1.644750924s, 5 calls | j2: 0.395%, 1.072519408s, 4 calls
j1: 0.584%, 1.644750924s, 5 calls | j2: 0.416%, 1.173230030s, 4 calls
j1: 0.564%, 1.644750924s, 5 calls | j2: 0.436%, 1.273398048s, 4 calls
j1: 0.545%, 1.644750924s, 5 calls | j2: 0.455%, 1.374261959s, 4 calls
j1: 0.527%, 1.644750924s, 5 calls | j2: 0.473%, 1.474492000s, 4 calls
j1: 0.527%, 1.698508948s, 5 calls | j2: 0.473%, 1.521447279s, 5 calls
```


CHANGELOG.md
# 0.1.1

* clean code

# 0.1.0

* initial version, all tests passed

calculator_test.go
package bottleneck

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestCalculator(t *testing.T) {
	c := NewCalculator()
	c.TimeSlice(0)
	time.Sleep(time.Millisecond * 100)
	c.TimeSlice(1)
	time.Sleep(time.Millisecond * 100)
	c.TimeSlice(0)
	time.Sleep(time.Millisecond * 100)

	entries := c.Stats()
	assert.EqualValues(t, 1, entries[0].CallCount)
	assert.EqualValues(t, 1, entries[1].CallCount)
	assert.InDelta(t, .666, entries[0].Percentage, .03)
	assert.InDelta(t, .333, entries[1].Percentage, .03)
	assert.InDelta(t, 200, int64(entries[0].Duration)/1000000, 10)
	assert.InDelta(t, 100, int64(entries[1].Duration)/1000000, 10)

	c.TimeSlice(0)

	entries = c.Stats()
	assert.EqualValues(t, 2, entries[0].CallCount)
	assert.EqualValues(t, 1, entries[1].CallCount)
	assert.InDelta(t, .666, entries[0].Percentage, .03)
	assert.InDelta(t, .333, entries[1].Percentage, .03)
	assert.InDelta(t, 200, int64(entries[0].Duration)/1000000, 10)
	assert.InDelta(t, 100, int64(entries[1].Duration)/1000000, 10)
}


calculator.go
package bottleneck

import (
	"sync/atomic"
	"time"
)

// Entry describes a bottleneck.
type Entry struct {
	Duration   time.Duration
	CallCount  int64
	Percentage float64
}

// Index describes a concrete bottleneck entry.
type Index int

// Possible Index values.
const (
	Index0 Index = iota
	Index1
	Index2
	Index3
	Index4
	Index5
	Index6
	Index7
	IndexCount
)

// Calculator allows to find bottlenecks in your code.
type Calculator struct {
	entries []Entry
	last    uint64
}

// NewCalculator creates an instance of Calculator.
func NewCalculator() *Calculator {
	return &Calculator{make([]Entry, int(IndexCount)), 0}
}

// TimeSlice starts a new time slice and adds current time slice to
// the bottleneck specified by the index in the previous call.
// Must be always called from the same goroutine.
func (c *Calculator) TimeSlice(index Index) {
	n := uint64(time.Now().UnixNano())
	last := atomic.SwapUint64(&c.last, n|uint64(index)<<61)
	if last != 0 {
		lastIndex := last >> 61
		last &^= 7 << 61
		atomic.AddInt64((*int64)(&c.entries[lastIndex].Duration), int64(n-last))
		atomic.AddInt64(&c.entries[lastIndex].CallCount, 1)
	}
}

// Stats returns bottleneck entries.
// Could be called from the different goroutines.
func (c *Calculator) Stats() []Entry {
	n := uint64(time.Now().UnixNano())
	for {
		last := atomic.LoadUint64(&c.last)
		if last == 0 {
			break
		}
		index := last >> 61
		if atomic.CompareAndSwapUint64(&c.last, last, n|index<<61) {
			atomic.AddInt64((*int64)(&c.entries[index].Duration), int64(n-last&^(7<<61)))
			break
		}
	}

	var total time.Duration
	result := make([]Entry, len(c.entries))
	for i := 0; i < len(c.entries); i++ {
		result[i].Duration = time.Duration(atomic.LoadInt64((*int64)(&c.entries[i].Duration)))
		result[i].CallCount = atomic.LoadInt64(&c.entries[i].CallCount)
		total += result[i].Duration
	}
	for i := 0; i < len(c.entries); i++ {
		result[i].Percentage = float64(result[i].Duration) / float64(total)
	}
	return result
}


examples/monitor/main.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"

	"github.com/ssgreg/bottleneck"
)

func main() {
	wg := sync.WaitGroup{}
	wg.Add(1)

	// Our bc has 3 bottleneck entries:
	// 0 - auxiliary: for loop, checks if break is needed
	// 1 - complex job (1)
	// 2 - complex job (2)
	bc := bottleneck.NewCalculator()

	go func() {
		defer wg.Done()
		finish := time.Now().Add(time.Second * 50)
		for {
			if time.Now().After(finish) {
				break
			}

			// Next time slice will be added to the bottleneck with index 1.
			bc.TimeSlice(bottleneck.Index1)
			// Do complex job (1).
			time.Sleep(time.Duration(rand.Int() % 500 * 1000000))

			// Next time slice will be added to the bottleneck with index 2.
			bc.TimeSlice(bottleneck.Index2)
			// Do complex job (2). It's 2 times longer than job (1).
			time.Sleep(time.Duration(rand.Int() % 1000 * 1000000))

			// Next time slice will be added to the bottleneck with index 0.
			bc.TimeSlice(bottleneck.Index0)
		}
	}()

	go func() {
		// Monitor that 10 times per second prints bottleneck stats.
		for {
			time.Sleep(time.Millisecond * 100)
			bns := bc.Stats()
			fmt.Printf("w1: %0.3f%%, %v, %d calls | w2: %0.3f%%, %v, %d calls\n", bns[1].Percentage, bns[1].Duration, bns[1].CallCount, bns[2].Percentage, bns[2].Duration, bns[2].CallCount)
		}
	}()

	wg.Wait()
}


doc.go
/*
Package bottleneck allows do detect bottlenecks in your code. Package uses atomic to minimize impact on performance. Bottleneck info returning by `Stats` function is eventually consistent.

Which operation takes more time?

The example below proves that creating and filling an array with random numbers takes less time than sorting.

    for i := 1000; i < 1000000000; i *= 10 {
        // Our bc has 2 bottleneck entries:
        // 0 - array creation and filling with random numbers
        // 1 - sorting
        bc := bottleneck.NewCalculator()

        // Will add array creation and filling time slice to the
        // bottleneck entry (0).
        bc.TimeSlice(bottleneck.Index0)

        ra := make([]int, i)
        for j := 0; j < i; j++ {
            ra[j] = rand.Int()
        }

        // Will add sorting time slice to the bottleneck entry (1).
        bc.TimeSlice(bottleneck.Index1)

        sort.Ints(ra)

        entries := bc.Stats()
        fmt.Printf("array len is %v, creating new random array takes %v, sorting takes %v\n", i, entries[0].Duration, entries[1].Duration)
    }

The output is:

	array len is 1000, creating new random array takes 40.059µs, sorting takes 213.93µs
	array len is 10000, creating new random array takes 455.646µs, sorting takes 2.163691ms
	array len is 100000, creating new random array takes 4.297062ms, sorting takes 26.020184ms
	array len is 1000000, creating new random array takes 40.96832ms, sorting takes 307.671125ms
	array len is 10000000, creating new random array takes 390.894645ms, sorting takes 3.574017192s
	array len is 100000000, creating new random array takes 3.911881876s, sorting takes 40.516029548s

Monitoring of a goroutine execution

The example below starts two goroutines. The first one is imitating a worker with two complex jobs. The second one monitors worker's executions. It shows percentage, number of calls and total execution time per each job.

	func main() {
		wg := sync.WaitGroup{}
		wg.Add(1)

		// Our bc has 3 bottleneck entries:
		// 0 - auxiliary: for loop, checks if break is needed
		// 1 - complex job (1)
		// 2 - complex job (2)
		bc := bottleneck.NewCalculator()

		go func() {
			defer wg.Done()
			finish := time.Now().Add(time.Second * 50)
			for {
				if time.Now().After(finish) {
					break
				}

				// Next time slice will be added to the bottleneck with index 1.
				bc.TimeSlice(bottleneck.Index1)

				// Do complex job (1).
				time.Sleep(time.Duration(rand.Int() % 500 * 1000000))
				//

				// Next time slice will be added to the bottleneck with index 2.
				bc.TimeSlice(bottleneck.Index2)

				// Do complex job (2). It's 2 times longer than job (1).
				time.Sleep(time.Duration(rand.Int() % 1000 * 1000000))
				//

				// Next time slice will be added to the bottleneck with index 0.
				bc.TimeSlice(bottleneck.Index0)
			}
		}()

		go func() {
			// Monitor that 10 times per second prints bottleneck stats.
			for {
				time.Sleep(time.Millisecond * 100)
				bns := bc.Stats()
				fmt.Printf("j1: %0.3f%%, %v, %d calls | j2: %0.3f%%, %v, %d calls\n", bns[1].Percentage, bns[1].Duration, bns[1].CallCount, bns[2].Percentage, bns[2].Duration, bns[2].CallCount)
			}
		}()

		wg.Wait()
	}

A fragment of output:

	j1: 1.000%, 100.260332ms, 0 calls | j2: 0.000%, 0s, 0 calls
	j1: 1.000%, 201.013773ms, 0 calls | j2: 0.000%, 0s, 0 calls
	j1: 1.000%, 301.297164ms, 0 calls | j2: 0.000%, 0s, 0 calls
	j1: 1.000%, 402.127131ms, 0 calls | j2: 0.000%, 0s, 0 calls
	j1: 0.815%, 410.234664ms, 1 calls | j2: 0.185%, 92.8835350ms, 0 calls
	j1: 0.679%, 410.234664ms, 1 calls | j2: 0.321%, 193.718235ms, 0 calls
	j1: 0.582%, 410.234664ms, 1 calls | j2: 0.418%, 294.723694ms, 0 calls
	j1: 0.509%, 410.234664ms, 1 calls | j2: 0.491%, 395.139638ms, 0 calls
	j1: 0.453%, 410.234664ms, 1 calls | j2: 0.547%, 495.800695ms, 0 calls
	j1: 0.452%, 454.419565ms, 1 calls | j2: 0.548%, 551.816673ms, 1 calls
	j1: 0.502%, 555.145899ms, 1 calls | j2: 0.498%, 551.816673ms, 1 calls
	j1: 0.543%, 655.362978ms, 1 calls | j2: 0.457%, 551.816673ms, 1 calls
	j1: 0.559%, 731.396896ms, 2 calls | j2: 0.441%, 576.023745ms, 1 calls
	j1: 0.571%, 804.403279ms, 2 calls | j2: 0.429%, 603.695591ms, 2 calls
	j1: 0.600%, 905.306768ms, 2 calls | j2: 0.400%, 603.695591ms, 2 calls
	j1: 0.625%, 1.005623715s, 2 calls | j2: 0.375%, 603.695591ms, 2 calls
	j1: 0.647%, 1.106392719s, 2 calls | j2: 0.353%, 603.695591ms, 2 calls
	j1: 0.646%, 1.169336820s, 3 calls | j2: 0.354%, 641.748760ms, 2 calls
	j1: 0.612%, 1.169336820s, 3 calls | j2: 0.388%, 741.900221ms, 2 calls
	j1: 0.581%, 1.169336820s, 3 calls | j2: 0.419%, 842.655883ms, 2 calls
	j1: 0.563%, 1.188210892s, 3 calls | j2: 0.437%, 923.956385ms, 3 calls
	j1: 0.582%, 1.289025414s, 3 calls | j2: 0.418%, 923.956385ms, 3 calls
	j1: 0.601%, 1.390096321s, 3 calls | j2: 0.399%, 923.956385ms, 3 calls
	j1: 0.591%, 1.428142697s, 4 calls | j2: 0.409%, 986.964089ms, 3 calls
	j1: 0.574%, 1.443765872s, 4 calls | j2: 0.426%, 1.072230598s, 4 calls
	j1: 0.590%, 1.544728063s, 4 calls | j2: 0.410%, 1.072230598s, 4 calls
	j1: 0.605%, 1.644750924s, 5 calls | j2: 0.395%, 1.072519408s, 4 calls
	j1: 0.584%, 1.644750924s, 5 calls | j2: 0.416%, 1.173230030s, 4 calls
	j1: 0.564%, 1.644750924s, 5 calls | j2: 0.436%, 1.273398048s, 4 calls
	j1: 0.545%, 1.644750924s, 5 calls | j2: 0.455%, 1.374261959s, 4 calls
	j1: 0.527%, 1.644750924s, 5 calls | j2: 0.473%, 1.474492000s, 4 calls
	j1: 0.527%, 1.698508948s, 5 calls | j2: 0.473%, 1.521447279s, 5 calls
*/
package bottleneck
